### ISSUE TITLE
The default value for a field in an entity

### ISSUE THREAD

**Issue by LookBad** [JetBrains/Exposed#1071]
Hi!

How can I set a default value for the field in an entity?
Function "getNextNumber" represents default value for number field.

```
    var number by ScenarioKeys.number

    fun getNextNumber() =
        find { (ScenarioKeys.product eq product.id) and (ScenarioKeys.scenario eq scenario.id) }
            .maxByOrNull { number }?.number ?: 1
```
I cannot find any information about it in the documentation.

**Comment by LookBad** [comment]
maxByOrNull and maxOfOrNull return java NPE :|

**Comment by Tapac** [comment]
Could you please share a bit more information? How you declare the default value, what result do you expect?

**Comment by LookBad** [comment]
```
class ScenarioKeyEntity(id: EntityID<Int>) : BaseIntEntity(id, ScenarioKeys) {
    companion object : BaseIntEntityClass<ScenarioKeyEntity>(ScenarioKeys)

    var product by ProductEntity referencedOn ScenarioKeys.product
    var scenario by ScenarioEntity referencedOn ScenarioKeys.scenario
    var number by ScenarioKeys.number
    var redirect by ScenarioKeyEntity optionalReferencedOn ScenarioKeys.redirect

    fun getNextNumber() =
        find { ScenarioKeys.product eq product.id and (ScenarioKeys.scenario eq scenario.id) }
            .maxOfOrNull { number }
            ?.plus(1) ?: 1
}
```

I want to create an entity with the number property which always set result from the "getNextNumber" method as value.
```

    override fun createScenario(scenarioParams: CreateScenarioParams) = transaction {
        val productContext = findProductOrThrowResourceNotFoundException(scenarioParams.productId)
        val scenarioContext = ScenarioEntity.new {
            name = scenarioParams.name
            description = scenarioParams.description
        }
        ScenarioKeyEntity.new {
            scenario = scenarioContext
            product = productContext
            number = getNextNumber()
        }
        scenarioContext.toPojo()
    }
```
Now I just call the "getNextNumber" method in the new block. I don't like this solution, it's inelegant.

**Comment by Tapac** [comment]
You could try to put it inside `clientDefault` on `ScenarioKeys.number` column. It will be initialized on any insert into ScenarioKeys table.

**Comment by luk-bed** [comment]
Ok, but in the `getNextNumber` method I run a query that requires the input context. In `clientDefault` I don't have access to this.

**Comment by LookBad** [comment]
@Tapac I can't use `clientDefault` :( Do you know anything about NPE in maxOfOrNull?

**Comment by Tapac** [comment]
@LookBad , no, I'm not aware of such problems. If you don't use !! it shouldn't throw NPE.

**Comment by LookBad** [comment]
Ok, so I found a bug :(

**Comment by LookBad** [comment]
@Tapac Do you have a roadmap for this project? Can you add bug label to this issue?

**Comment by Tapac** [comment]
Could you share the whole project with a reproducible test as I'm not sure that NPE is related to Exposed

**Comment by LookBad** [comment]
Ok, I can share to you my project.
[Archive.zip](https://github.com/JetBrains/Exposed/files/5447237/Archive.zip)

**Comment by Tapac** [comment]
If understand properly you are trying to call getNextNumber() function on entity which wasn't created yet.
I would advice to make it function of `ProductEntity`, replace it with max sql function, and call it like:
```kotlin
class ProductEntity() ... {
   fun getNextNumber() : Int {
       val maxNumber = ScenarioKeys.number.max() 
       return ScenarioKeys.slice(maxNumber)
             .select { ScenarioKeys.product eq this.id }
             .singleOrNull()?.tryGet(maxNumber)?.plus(1) ?: 1
   }
}

fun getNewScenarioKey(product: ProductEntity, scenario: ScenarioEntity, redirect: ScenarioKeyEntity? = null) =
    ScenarioKeyEntity.new {
        this.product = product
        this.number = product.getNextNumber()
        this.scenario = scenario
        this.redirect = redirect
    }


```

**Comment by LookBad** [comment]
@Tapac Ok, I can do it but it doesn't look like a legit solution. The `getNextNumber` should be a function in a db triggered on insert action - but it isn't possible to implement now, am I right?

**Comment by Tapac** [comment]
Right now there is no support for defining triggers with Exposed but you can write it with raw SQL and execute for your database.

**Comment by LookBad** [comment]
Ok, I decided to move settings number to trigger, but now I have to set a default value or I get:
`BatchDataInconsistentException: Can't add a new batch because columns: scenariokey."number" don't have default values. DB defaults don't support in batch inserts`. 
If I set "default(0)" or "clientDefault { 0 }" insert works correctly, but Entity uses values provided by "default" and ignore values provided by the database. 
How can I fix it?
